Added Support for Touch devices

chinmaydalvi opened the pull request:
# Merge Checklist
## :wrench: Issue(s) fixed:
- [ ] Fixes https://github.com/omnisci/mapd-draw/issues/28

## :smoking: Smoke Test
- [ ] Works in chrome
- [ ] Works in firefox
- [ ] Works in safari

## :ship: Merge
- [ ] author crafted PR's title into release-worthy commit message.

jrajav commented:
Hi @chinmaydalvi , thank you for the contribution!

Did this work for you on a touch device? It seems to only drag the map around, rather than draw the lasso shape, when I try it hooked into mapd-charting, on the example you mentioned in https://github.com/omnisci/mapd-draw/issues/28: https://omnisci.github.io/mapd-charting/example/example5.html

chinmaydalvi commented:
Hi @jrajav 
Thanks for your quick reply and sorry for the confusion.

In the PR I have added the missing touch devices events for MapD Draw Library. 
mapd-draw library gives Basic functionality to place a shape on the page and drag that shape anywhere on the page as well as resize the shape.
Refer the following example in mapd-draw Library and check this on touch device
http://127.0.0.1:8080/example/exampleShapeBuilder.html

Above example is working perfectly fine in case of Nontouch devices as all mouse pointer events are present already in the Library. But on touch devices, the above example is not working as touch events are missing in Library.

In this PR I have added those missing touch devices events in Library so that the above example can work in case of touch devices.


Regarding your query related to Lasso Tool example 
https://omnisci.github.io/mapd-charting/example/example5.html

Lasso Tool is created on the top of mapd-draw Library. In mapd charting library, custom code of your Lasso tool example is written.
This custom code of Lasso tool is also not supported on a touch device, therefore, Lasso example is not working on touch devices.

To support Lasso Tool on touch devices we need to add support in custom code of Lasso example too which is there in mapd charting Library **but Before that, we need to add Basic touch events support in Mapd draw library as without these changes in mapd draw library we can't directly provide support in Lasso's custom code** 

Therefore In this Pull request, I have added the support Mapd draw Library first. 

I hope that you would understand

Thanks,
Chinmay Dalvi

chinmaydalvi commented:
Hi @jrajav 
Any doubt/question related to this comment https://github.com/omnisci/mapd-draw/pull/29#issuecomment-515699555 ?

vastcharade commented on src/engine/shape-builder.js:
```diff
@@ -338,11 +394,34 @@ export default class ShapeBuilder extends DrawEngine {
 
   _mousemoveCB(event) {
+    this.setDenyMouseEventFlag(event)
+    if (this.denyMouseEvent && !event.touches) {
```
there's a lot of checks for `event.touches` since you've collapsed the mouse/touch callbacks together. I think it'd be a bit more readable to turn this into a utility function: `isTouchEvent(event)`. It would return true if the event was a touch event.

vastcharade commented on src/engine/shape-builder.js:
```diff
@@ -338,11 +394,34 @@ export default class ShapeBuilder extends DrawEngine {
+        event.preventDefault()
+      }
+      if ((Date.now() - this.firstTapTime) < DOUBLE_CLICK_DELAY) {
+        this._dblclickCB(event)
```
This looks ok. Did you verify that this works like a mouse double-click event. In particular I'm wondering if a mouse double-click event actually runs the mousedown/mouseup twice first before firing the double click event. If that's true, then this looks right because the touchstart/touchend would be fired twice before `_dblclickCB` is called.
